src/mesh/amr_loadbalance.cpp

```diff
@@ -462,368 +697,183 @@ void Mesh::RedistributeAndRefineMeshBlocks(ParameterInput *pin, ApplicationInput
-            if (ranklist[on + ll] != Globals::my_rank) continue;
-            // fine to coarse on the same MPI rank (different AMR level) - restriction
-            auto pob = FindMeshBlock(on + ll);
-
-            // allocate sparse variables that were allocated on old block
-            for (auto var : pob->meshblock_data.Get()->GetVariableVector()) {
-              if (var->IsSparse() && var->IsAllocated()) {
-                new_block_list[n - nbs]->AllocateSparse(var->label());
-              }
-            }
-            FillSameRankFineToCoarseAMR(pob.get(), new_block_list[n - nbs].get(),
-                                        loclist[on + ll]);
-          }
-        } else if ((loclist[on].level < newloc[n].level) && // coarse to fine (c2f)
-                   (ranklist[on] == Globals::my_rank)) {
-          // coarse to fine on the same MPI rank (different AMR level) - prolongation
-          auto pob = FindMeshBlock(on);
-
-          // allocate sparse variables that were allocated on old block
-          for (auto var : pob->meshblock_data.Get()->GetVariableVector()) {
-            if (var->IsSparse() && var->IsAllocated()) {
-              new_block_list[n - nbs]->AllocateSparse(var->label());
-            }
-          }
-          FillSameRankCoarseToFineAMR(pob.get(), new_block_list[n - nbs].get(),
-                                      newloc[n]);
-        }
       }
+    } else {
+      // on a different refinement level or MPI rank - create a new block
+      BoundaryFlag block_bcs[6];
+      SetBlockSizeAndBoundaries(newloc[n], block_size, block_bcs);
+      // append new block to list of MeshBlocks
+      new_block_list[n - nbs] =
+          MeshBlock::Make(n, n - nbs, newloc[n], block_size, block_bcs, this, pin, app_in,
+                          packages, resolved_packages, gflag);
     }
+  }
 
-    // Replace the MeshBlock list
-    block_list = std::move(new_block_list);
+  // Replace the MeshBlock list
+  auto old_block_list = std::move(block_list);
+  block_list = std::move(new_block_list);
 
-    // Ensure local and global ids are correct
-    for (int n = nbs; n <= nbe; n++) {
-      block_list[n - nbs]->gid = n;
-      block_list[n - nbs]->lid = n - nbs;
-    }
+  // Ensure local and global ids are correct
+  for (int n = nbs; n <= nbe; n++) {
+    block_list[n - nbs]->gid = n;
+    block_list[n - nbs]->lid = n - nbs;
   }
+
   Kokkos::Profiling::popRegion(); // Step 8: Construct new MeshBlockList
 
   // Step 9. Receive the data and load into MeshBlocks
   Kokkos::Profiling::pushRegion("Step 9: Recv data and unpack");
-  // This is a test: try MPI_Waitall later.
-#ifdef MPI_PARALLEL
-  if (nrecv != 0) {
-    int test;
-    std::vector<bool> received(nrecv, false);
-    int rb_idx;
+  bool all_received;
+  int niter = 0;
+  if (block_list.size() > 0) {
+    std::vector<bool> finished(
+        std::max((nbe - nbs + 1), 1) * FindMeshBlock(nbs)->vars_cc_.size() * 8, false);
```

Comment:
Add comment? (It's not immediately obvious to me where the `>vars_cc_.size() * 8` comes from.

Reply:
This is just to make sure that this array has enough space in the worst case scenario (where every block on the rank is a new coarser block so that it is receiving a message from eight separate finer blocks for each variable). I have added a comment to this effect since I also forgot where the magic 8 comes from.